src/validation/global_validator.rs

```diff
@@ -0,0 +1,113 @@
+use crate::{
+    ast::SourceRange,
+    diagnostics::Diagnostic,
+    index::{symbol::SymbolMap, Index},
+};
+
+/// Validator that does not check a dedicated file but rather
+/// uses the index to validate the project as a whole.
+/// It performs validations including:
+///  - naming-conflicts
+///  - <tbc>
+pub struct GlobalValidator {
+    pub diagnostics: Vec<Diagnostic>,
+}
+
+impl GlobalValidator {
+    pub fn new() -> GlobalValidator {
+        GlobalValidator {
+            diagnostics: Vec::new(),
+        }
+    }
+
+    /// reports a name-conflict for the given name. the locations indicate the
+    /// locations of the declared symbols that make up the conflict. this method will
+    /// create a diagnostic per location where it attaches the other locations as additional information.
+    fn report_name_conflict(&mut self, name: &str, locations: &[&SourceRange]) {
```

Comment:
You can have a generic iterator here : 
`report_name_conflict(&mut self, name: &str, locations: T) where T : Iterator<Item = SourceRange>` to avoid the collects below
Didn't try it, but I think since the iterator type is known at compile time this should work 
Got it from here : https://stackoverflow.com/questions/34969902/how-to-write-a-rust-function-that-takes-an-iterator

Reply:
yes i could pass an iterator here, but I need to iterate the thing several times: 
i report like that:

```
for l in locations {
   others = locations.filter(it !== l);
   report(l, others);
}
````
you have an idea how i can do this with one iterator?

Reply:
No, didn't notice this. I think slice works best here in that case.